crates/pop-contracts/src/up.rs

```diff
@@ -93,9 +102,150 @@ pub async fn set_up_upload(
+			instantiate_exec.args().value(),
+			gas_limit,
+			storage_deposit_limit,
+			hash.clone(),
+			instantiate_exec.args().data().into(),
+			instantiate_exec.args().salt().into(),
+		)
+		.build()
+		.encode_call_data_to(&instantiate_exec.client().metadata(), &mut encoded_data),
+	}?;
+
+	Ok(encoded_data)
+}
+
+pub async fn get_contract_code(
+	path: Option<&PathBuf>,
+) -> anyhow::Result<contract_extrinsics::WasmCode> {
+	let manifest_path = get_manifest_path(path.map(|p| p as &Path))?;
+
+	// signer does not matter for this
+	let signer = create_signer("//Alice")?;
+	let extrinsic_opts =
+		ExtrinsicOptsBuilder::<DefaultConfig, DefaultEnvironment, Keypair>::new(signer)
+			.manifest_path(Some(manifest_path))
+			.done();
+	let artifacts = extrinsic_opts.contract_artifacts()?;
+
+	let artifacts_path = artifacts.artifact_path().to_path_buf();
+	let code = artifacts.code.ok_or_else(|| {
+		Error::UploadContractError(format!(
+			"Contract code not found from artifact file {}",
+			artifacts_path.display()
+		))
+	})?;
+	Ok(code)
+}
+
+pub async fn upload_contract_signed(
+	url: &str,
+	payload: String,
+) -> anyhow::Result<UploadResult<SubstrateConfig>> {
+	let events = submit_signed_payload(url, payload).await?;
+
+	let code_stored = events.find_first::<CodeStored<subxt::config::substrate::H256>>()?;
+
+	Ok(UploadResult { code_stored, events })
+}
+
+pub async fn instantiate_contract_signed(
+	url: &str,
+	payload: String,
+) -> anyhow::Result<InstantiateExecResult<SubstrateConfig>> {
+	let events = submit_signed_payload(url, payload).await?;
+
+	// The CodeStored event is only raised if the contract has not already been
+	// uploaded.
+	let code_hash = events
+		.find_first::<CodeStored<subxt::config::substrate::H256>>()?
+		.map(|code_stored| code_stored.code_hash);
+
+	let instantiated = events
+		.find_first::<ContractInstantiated<subxt::config::substrate::AccountId32>>()?
+		.ok_or_else(|| {
+			Error::InstantiateContractError("Failed to find Instantiated event".to_string())
+		})?;
+
+	Ok(InstantiateExecResult { events, code_hash, contract_address: instantiated.contract })
+}
+
+pub async fn submit_signed_payload(
+	url: &str,
+	payload: String,
+) -> anyhow::Result<ExtrinsicEvents<SubstrateConfig>> {
+	let rpc_client = subxt::backend::rpc::RpcClient::from_url(url).await?;
+	let client = subxt::OnlineClient::<SubstrateConfig>::from_rpc_client(rpc_client).await?;
+
+	let hex_encoded = from_hex(&payload)?;
+
+	let extrinsic = SubmittableExtrinsic::from_bytes(client, hex_encoded);
+
+	// src: https://github.com/use-ink/cargo-contract/blob/68691b9b6cdb7c6ec52ea441b3dc31fcb1ce08e0/crates/extrinsics/src/lib.rs#L143
+
+	use subxt::{
+		error::{RpcError, TransactionError},
+		tx::TxStatus,
+	};
+
+	let mut tx = extrinsic.submit_and_watch().await?;
+
+	while let Some(status) = tx.next().await {
+		match status? {
+			TxStatus::InBestBlock(tx_in_block) | TxStatus::InFinalizedBlock(tx_in_block) => {
```

Comment:
Just to be aware for `pop call chain` we wait until is finalized. 

Does it make sense to use the same approach  as `cargo_contracts`?. Based on the comment `cargo_contract` has here: https://github.com/use-ink/cargo-contract/blob/68691b9b6cdb7c6ec52ea441b3dc31fcb1ce08e0/crates/extrinsics/src/lib.rs#L171 this code is specific for `substrate-contracts-node`, but are there any potential use cases for secure signing on the testing chain `substrate-contracts-node`?